pkg/p2p/server.go

```diff
@@ -0,0 +1,835 @@
+			// rare and indicates problems with the receiver's handler.
+			// It is expected to happen only with extreme system latency or buggy code.
+			//
+			// Reports an error so that the receiver can gracefully exit.
+			return cerror.ErrPeerMessageDataLost.GenWithStackByArgs(entry.Topic, lastAck+1)
+		}
+
+		if err := unmarshalMessage(entry.Content, e); err != nil {
+			return cerror.WrapError(cerror.ErrPeerMessageDecodeError, err)
+		}
+
+		if err := fn(sm.SenderId, e); err != nil {
+			return errors.Trace(err)
+		}
+
+		m.acks.Set(sm.SenderId, entry.GetTopic(), entry.GetSequence())
+
+		return nil
+	}).OnExit(func(err error) {
+		log.Warn("topic handler returned error", zap.Error(err))
+		_ = m.scheduleTask(ctx, taskOnDeregisterHandler{
+			topic: topic,
+		})
+	})
+
+	doneCh := make(chan struct{})
+
+	if err := m.scheduleTask(ctx, taskOnRegisterHandler{
+		topic:   topic,
+		handler: poolHandle,
+		done:    doneCh,
+	}); err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	return doneCh, poolHandle.ErrCh(), nil
+}
+
+// SyncRemoveHandler removes the registered handler for the given topic and wait
+// for the operation to complete.
+func (m *MessageServer) SyncRemoveHandler(ctx context.Context, topic string) error {
+	doneCh, err := m.RemoveHandler(ctx, topic)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case <-doneCh:
+	case <-m.closeCh:
+		log.Debug("message server is closed while a handler is being removed",
+			zap.String("topic", topic))
+		return nil
+	}
+
+	return nil
+}
+
+// RemoveHandler removes the registered handler for the given topic.
+func (m *MessageServer) RemoveHandler(ctx context.Context, topic string) (chan struct{}, error) {
+	doneCh := make(chan struct{})
+	if err := m.scheduleTask(ctx, taskOnDeregisterHandler{
+		topic: topic,
+		done:  doneCh,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return doneCh, nil
+}
+
+func (m *MessageServer) registerHandler(ctx context.Context, topic string, handler workerpool.EventHandle, doneCh chan struct{}) error {
+	defer close(doneCh)
+
+	if _, ok := m.handlers[topic]; ok {
+		// allow replacing the handler here would result in behaviors difficult to define.
+		// Continuing the program when there is a risk of duplicate handlers will likely
+		// result in undefined behaviors, so we panic here.
+		log.Panic("duplicate handlers",
+			zap.String("topic", topic))
+	}
+
+	m.handlers[topic] = handler
+	if err := m.handlePendingMessages(ctx, topic); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// handlePendingMessages must be called with `handlerLock` taken exclusively.
+func (m *MessageServer) handlePendingMessages(ctx context.Context, topic string) error {
+	for key, entries := range m.pendingMessages {
+		if key.Topic != topic {
+			continue
+		}
+
+		for _, entry := range entries {
+			if err := m.handleMessage(ctx, entry.StreamMeta, entry.Entry); err != nil {
```

Comment:
Are they covered by tests?

Reply:
They were to be covered by the `server_client_integration_test.go`. I have modified the unit test case so that these lines are covered.